test-utils/src/dev/seed.rs

```diff
@@ -0,0 +1,228 @@
+
+use std::io::{BufWriter, Write};
+
+use anyhow::{ensure, Context, Result};
+use camino::{Utf8Path, Utf8PathBuf};
+use slog::Logger;
+
+/// The environment variable via which the path to the seed tarball is passed.
+pub static CRDB_SEED_TAR_ENV: &str = "CRDB_SEED_TAR";
+
+/// Creates a string identifier for the current DB schema and version.
+//
+/// The goal here is to allow to create different "seed" tarballs
+/// for each revision of the DB.
+pub fn digest_unique_to_schema() -> String {
+    let schema = include_str!("../../../schema/crdb/dbinit.sql");
+    let crdb_version = include_str!("../../../tools/cockroachdb_version");
+    let mut ctx = ring::digest::Context::new(&ring::digest::SHA256);
+    ctx.update(&schema.as_bytes());
+    ctx.update(&crdb_version.as_bytes());
+    let digest = ctx.finish();
+    hex::encode(digest.as_ref())
+}
+
+/// Looks up the standard environment variable `CRDB_SEED_INVALIDATE` to check
+/// if a seed should be invalidated. Returns a string to pass in as the
+/// `why_invalidate` argument of [`ensure_seed_tarball_exists`].
+pub fn should_invalidate_seed() -> Option<&'static str> {
+    (std::env::var("CRDB_SEED_INVALIDATE").as_deref() == Ok("1"))
+        .then_some("CRDB_SEED_INVALIDATE=1 set in environment")
+}
+
+/// The return value of [`ensure_seed_tarball_exists`].
+#[derive(Clone, Copy, Debug)]
+pub enum SeedTarballStatus {
+    Created,
+    Invalidated,
+    Existing,
+}
+
+impl SeedTarballStatus {
+    pub fn log(self, log: &Logger, seed_tar: &Utf8Path) {
+        match self {
+            SeedTarballStatus::Created => {
+                info!(log, "Created CRDB seed tarball: `{seed_tar}`");
+            }
+            SeedTarballStatus::Invalidated => {
+                info!(
+                    log,
+                    "Invalidated and created new CRDB seed tarball: `{seed_tar}`",
+                );
+            }
+            SeedTarballStatus::Existing => {
+                info!(log, "Using existing CRDB seed tarball: `{seed_tar}`");
+            }
+        }
+    }
+}
+
+/// Ensures that a seed tarball corresponding to the schema returned by
+/// [`digest_unique_to_schema`] exists, recreating it if necessary.
+///
+/// If `why_invalidate` is `Some`, then if the seed tarball exists, it will be
+/// deleted before being recreated.
+///
+/// # Notes
+///
+/// This method should _not_ be used by tests. Instead, rely on the `crdb-seed`
+/// setup script.
+pub async fn ensure_seed_tarball_exists(
+    log: &Logger,
+    why_invalidate: Option<&str>,
+) -> Result<(Utf8PathBuf, SeedTarballStatus)> {
+    // If the CRDB_SEED_TAR_ENV variable is set, return an error.
+    if let Ok(val) = std::env::var(CRDB_SEED_TAR_ENV) {
+        anyhow::bail!(
+            "{CRDB_SEED_TAR_ENV} is set to `{val}` -- implying that a test called \
+             ensure_seed_tarball_exists. Instead, tests should rely on the `crdb-seed` \
+             setup script."
+        );
+    }
```

Comment:
This seems like a bit of a proxy for `#[ifcfg(not(test))]` -- is that too extreme of a conditional qualifier to put on this method?

It does seem like it would make `crdb-seed`, the binary, untestable, but I think it would enforce more accurately what the comments are describing.

Reply:
This is a dependency (not a leaf node) so I don't think `cfg(not(test))` would work? And if I try gating it behind a feature flag, the issue would be that `omicron-dev` needs this function and so feature unification would mean this gets included in test binaries if you're doing a top-level `cargo test`.

Reply:
Actually, I ended up both preserving this check and gating it behind a feature flag. So we're triply sure this code isn't used by another other than omicron-dev and crdb-seed.